crates/extensions/c8y_mapper_ext/src/availability/actor.rs

```diff
@@ -0,0 +1,295 @@
+        );
+
+        self.send_smartrest_set_required_availability_for_main_device()
+            .await?;
+        self.start_heartbeat_timer_if_interval_is_positive(&topic_id)
+            .await?;
+
+        Ok(())
+    }
+
+    async fn process_registration_message(
+        &mut self,
+        message: &EntityRegistrationMessage,
+    ) -> Result<(), RuntimeError> {
+        let source = &message.topic_id;
+
+        match message.r#type {
+            EntityType::MainDevice => match self.update_device_service_pair(message) {
+                RegistrationResult::New | RegistrationResult::Update => {
+                    self.start_heartbeat_timer_if_interval_is_positive(source)
+                        .await?;
+                }
+                RegistrationResult::Error(reason) => {
+                    warn!(reason)
+                }
+            },
+            EntityType::ChildDevice => match self.update_device_service_pair(message) {
+                RegistrationResult::New => {
+                    self.send_smartrest_set_required_availability_for_child_device(source)
+                        .await?;
+                    self.start_heartbeat_timer_if_interval_is_positive(source)
+                        .await?;
+                }
+                RegistrationResult::Update => {
+                    self.start_heartbeat_timer_if_interval_is_positive(source)
+                        .await?;
+                }
+                RegistrationResult::Error(reason) => warn!(reason),
+            },
+            EntityType::Service => {}
+        }
+
+        Ok(())
+    }
+
+    /// Insert a <"device topic ID" - "service topic ID" and "external ID"> pair into the map.
+    /// If @health is provided in the registration message, use the value as long as it's valid as a service topic ID.
+    /// If @health is not provided, use the "tedge-agent" service topic ID as default.
+    /// @id is the only source to know the device's external ID. Hence, @id must be provided in the registration message.
+    fn update_device_service_pair(
+        &mut self,
+        registration_message: &EntityRegistrationMessage,
+    ) -> RegistrationResult {
+        let source = &registration_message.topic_id;
+
+        let result = match registration_message.other.get("@health") {
+            None => {
+                Ok(registration_message.topic_id.to_default_service_topic_id("tedge-agent").unwrap())
+            }
+            Some(raw_value) => {
+                match raw_value.as_str() {
+                    None => Err(format!("'@health' must hold a string value. Given: {raw_value:?}")),
+                    Some(maybe_service_topic_id) => {
+                        EntityTopicId::from_str(maybe_service_topic_id)
+                            .map(|id| id.into())
+                            .map_err(|_| format!("'@health' must be the default service topic schema 'device/DEVICE_NAME/service/SERVICE_NAME'. Given: {maybe_service_topic_id}"))
```

Comment:
This error message is not aligned with what is checked: `maybe_service_topic_id` is expected to match  `EntityTopicId` and not necessarily a service id of the default MQTT scheme.

Reply:
I'm now thinking about the design.

Indeed, in terms of implementation level, it doesn't matter if the `@health` has either `EntityTopicId` or `ServiceTopicId`. `EntityTopicId` is just four segments, can be `custom/any/four/segs`. On the contrary, `ServiceTopicId` limits to the default scheme,`device/DEVICE_NAME/service/SERVICE_NAME`.

If `EntityTopicId`, the custom health endpoint can be `te/custom/any/four/segs/status/health` as well as the default service scheme.
If `ServiceTopicId`, the custom health endpoint can be only the default scheme, `te/device/DEVICE_NAME/service/SERVICE_NAME/status/health`.

Now I re-read the discussion in the initial PR, from the @reubenmiller 's [comment,](https://github.com/thin-edge/thin-edge.io/pull/2866#discussion_r1612930218), he said just entities, so I think I misunderstood that it needs to limit to `ServiceTopicId`. 

Any thoughts?

Reply:
I would accept any `te/+/+/+/+/status/health` messages.

If a registration message declares its `@heath` entity being `custom/any/four/segs`,
then simply uses it!

Reply:
Done in cb41d2df39dd6dfb81f47ee784af77e08368e2f6